Objects/genobject.c

```diff
@@ -408,9 +408,33 @@ gen_close(PyGenObject *gen, PyObject *args)
         PyErr_SetString(PyExc_RuntimeError, msg);
         return NULL;
     }
-    if (PyErr_ExceptionMatches(PyExc_StopIteration)
-        || PyErr_ExceptionMatches(PyExc_GeneratorExit)) {
-        PyErr_Clear();          /* ignore these errors */
+    if (PyErr_ExceptionMatches(PyExc_StopIteration)) {
+        /* retrieve the StopIteration exception instance being handled, and
+         * extract its value */
+        PyObject *exc, *args, *value;
+        PyThreadState *tstate = _PyThreadState_GET();
+        if (tstate == NULL) {
+            PyErr_Clear();
+            Py_RETURN_NONE;
+        }
+        exc = tstate->current_exception;
+        if (exc == NULL || !PyExceptionInstance_Check(exc)) {
+            PyErr_Clear();
```

Comment:
You can move this and later `PyErr_Clear()` calls up to the top (right after `PyErr_ExceptionMatches()`); none of the calls made here can set an exception.